### ISSUE TITLE
Windows 10 + WSL2 Not working

### ISSUE THREAD

**Issue by mhl787156** [Kalmat/PyWinCtl#71]
Hi! I'm currently building an application out of various python TK/QT/WX Guis (Don't ask...requirements...) and do a bunch of my dev on WSL2 on windows. PyWinCTL looks great for my application, but it's not quite working on WSL in particular I think. 

I have a basic script:
```python
import subprocess
import pywinctl as pwc
import time
command = ["python", "fleet_monitor.py", "example/basic_tee_spatially_separated_mission", "-n", "5"]
subprocess.Popen(command)
time.sleep(5)
titles = pwc.getAllTitles()
print(titles)
print(pwc.getAllWindows())
print(pwc.getAllAppsWindowsTitles())
```
which opens a TKinter window and the tries to find it. At which point it fails to find these windows and all the calls to pwc return empty.

I then cloned the repo and ran `pytest tests/test_pywinctl.py` - where it FAILED tests/test_pywinctl.py::test_basic - KeyError: 'XDG_CURRENT_DESKTOP'. Not sure what the correct entry is on WSL... but I set `export XDG_CURRENT_DESKTOP=ubuntu:GNOME` and that seemed to mollify it. 

Running the tests again I now get an error about Xlib which I'm not sure how to resolve which is below. 

```
=================================================================== test session starts ====================================================================
platform linux -- Python 3.9.16, pytest-7.4.0, pluggy-1.2.0
rootdir: /home/mickey/Portal/PyWinCtl
collected 1 item

tests/test_pywinctl.py F                                                                                                                             [100%]

========================================================================= FAILURES =========================================================================
________________________________________________________________________ test_basic ________________________________________________________________________

    def test_basic():

        if sys.platform == "win32":
            subprocess.Popen('notepad')
            time.sleep(0.5)

            testWindows = [pywinctl.getActiveWindow()]
            # testWindows = pywinctl.getWindowsWithTitle('Untitled - Notepad')   # Not working in other languages
            assert len(testWindows) == 1

            npw = testWindows[0]

            basic_win32(npw)

        elif sys.platform == "linux":
            subprocess.Popen('gedit')
            time.sleep(5)

            testWindows = [pywinctl.getActiveWindow()]
            assert len(testWindows) == 1

            npw = testWindows[0]

>           basic_linux(npw)

tests/test_pywinctl.py:37:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
tests/test_pywinctl.py:245: in basic_linux
    if npw.isMaximized:  # Make sure it starts un-maximized
../../mambaforge/envs/portal/lib/python3.9/site-packages/pywinctl/_pywinctl_linux.py:746: in isMaximized
    state = self._win.getWmState(True)
../../mambaforge/envs/portal/lib/python3.9/site-packages/ewmhlib/_ewmhlib.py:1324: in getWmState
    return getPropertyValue(self.getProperty(Window.WM_STATE), text, self.display)
../../mambaforge/envs/portal/lib/python3.9/site-packages/ewmhlib/_ewmhlib.py:1032: in getProperty
    return getProperty(self.xWindow, prop, prop_type, self.display)
../../mambaforge/envs/portal/lib/python3.9/site-packages/ewmhlib/_ewmhlib.py:180: in getProperty
    return window.get_full_property(prop, prop_type)
../../mambaforge/envs/portal/lib/python3.9/site-packages/Xlib/xobject/drawable.py:472: in get_full_property
    prop = self.get_property(property, property_type, 0, sizehint)
../../mambaforge/envs/portal/lib/python3.9/site-packages/Xlib/xobject/drawable.py:455: in get_property
    r = request.GetProperty(display = self.display,
../../mambaforge/envs/portal/lib/python3.9/site-packages/Xlib/protocol/rq.py:1368: in __init__
    self.reply()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <GetProperty serial = 19, data = None, error = <class 'Xlib.error.BadWindow'>: code = 3, resource_id = <Resource 0x00c00115>, sequence_number = 19, major_opcode = 20, minor_opcode = 0>

    def reply(self):
        # Send request and wait for reply if we hasn't
        # already got one.  This means that reply() can safely
        # be called more than one time.

        self._response_lock.acquire()
        while self._data is None and self._error is None:
            self._display.send_recv_lock.acquire()
            self._response_lock.release()

            self._display.send_and_recv(request = self._serial)
            self._response_lock.acquire()

        self._response_lock.release()
        self._display = None

        # If error has been set, raise it
        if self._error:
>           raise self._error
E           Xlib.error.BadWindow: <class 'Xlib.error.BadWindow'>: code = 3, resource_id = <Resource 0x00c00115>, sequence_number = 19, major_opcode = 20, minor_opcode = 0

../../mambaforge/envs/portal/lib/python3.9/site-packages/Xlib/protocol/rq.py:1388: BadWindow
================================================================= short test summary info ==================================================================
FAILED tests/test_pywinctl.py::test_basic - Xlib.error.BadWindow: <class 'Xlib.error.BadWindow'>: code = 3, resource_id = <Resource 0x00c00115>, sequence_number = 19, major_opcode = 20, minor_opc...
==================================================================== 1 failed in 5.34s =====================================================================
```

I'm assuming there's some libraries that WSL2 is missing or oddly configured by default which PyWinCtl might need? 

Thanks!

**Comment by Kalmat** [comment]
Wow! Interesting project and very challenging question!

PyWinCtl relies heavily on X11 and EWMH. Since WSL2 doesn't have any X-Server, it will not work. You would need to install a X-Server for Windows which, in addtion, is compliant with EWMH specs (don't even know if that exists). Even though you manage to do this, I am not sure it will fully work (I never tested myself, to be honest).

I am not familiar with WSL2, unfortunately. In a quick search, I found [this X-Server for Windows](https://medium.com/javarevisited/using-wsl-2-with-x-server-linux-on-windows-a372263533c3), which seems to have EWMH support thru XCB, as per [this](https://github.com/sasdf/vcxsrv/blob/master/include/xcb/xcb_ewmh.h), but just guessing.

Not sure if that all fits your needs or make any sense at all... Please let me know what you think.

Thank you!

**Comment by Kalmat** [comment]
Hi again!

Just guessing: did you check if those windows are included in pywinctl.getAllWindows()? I don't really think so but, if they do, you can manage them from Windows itself. If it works, you can take a look to [this](https://github.com/Kalmat/TidyDesk) which, if I understood you well, it is very similar to what you are looking for (though on "standard" Windows/Win32api, not Linux nor WSL2).

Sticking to WSL2/Linux, apart from X11, it is necessary EWMH. For instance, Wayland is not totally compliant with EWMH, so there is no way to get the active window nor the list of open windows (making it impossible to address just one window unless you already have its XID)... This drove me crazy for weeks until I eventually had to desist. According to the architecture you linked, it seems it is using Wayland as compositor. If so, you can give a try to this:

```
import json
import subprocess
from typing import List, Union

windowsList: List[dict[str, Union[str, bool]]] = [{}]
activeWindow: dict[str, Union[str, bool]] = {}
cmd = ('gdbus call --session --dest org.gnome.Shell --object-path /org/gnome/Shell '
       '--method org.gnome.Shell.Eval "global.get_window_actors()'
       '.map(a=>a.meta_window)'
       '.map(w=>({class: w.get_wm_class(), title: w.get_title(), active: w.has_focus(), id: w.get_description(), id2: w.get_id(), id3: w.get_pid()}))"')
ret = subprocess.check_output(cmd, shell=True, timeout=1).decode("utf-8").replace("\n", "")
if ret and ret.startswith("(true, "):
    windows: List[str] = (str(ret[8:-2]).replace("[", "").replace("]", "").replace("},{", "}|&|{").split("|&|"))
    for window in windows:
        output: dict[str, Union[str, bool]] = json.loads(window)
        print(output)
        windowsList.append(output)
        if output["active"]:
            activeWindow = output
```

Check if the target windows are printed using this script. If it complains about permissions, you will likely have to enable unsafe-mode... If so, it will not be usable for end-users (this is why I had to desist), but it can give you a good reference on what may be happening and how to address it.

Unfortunately, I am not aware of those tools you mention. I am quite new in all this about programming, and I code just as a hobby, sorry, so there are lots of tools and languages and environments I am not familiar with!

... A mess, I know... but, who said it was going to be easy? HAHAHAHA!

**Comment by Kalmat** [comment]
**EDIT**
I managed to install WSL2 on my system (just curious). Running pywinctl.getAllWindows() from Windows, it finds the Ubuntu window, and it can be targeted. Not sure if everything will properly work yet, but I was able to get and move the window from Windows itself.

![Captura](https://github.com/Kalmat/PyWinCtl/assets/39727793/a04b826c-1be1-4e9a-870d-8beba4b06a95)

Within WSL (Ubuntu), the `gdbus `command fails (most likely because of the unsafe-mode not enabled, but not sure).

Apart from this, do you have to stick to WSL? Or you can give a try to a VM on, e.g.,  Virtualbox, or even better install Ubuntu along with Windows (it can be done without affecting your whole system nor the boot procedure)?

**Comment by mhl787156** [comment]
Thanks for trying it out, yes I tried it over the weekend and the gdbus command fails for me too. I guess WSL passes off the creation and management of the window to the Windows systems. 

I could definitely do that, but they want a kinda cross-platform solution without VMs and such. We have one application which only works on Windows, and another only on Linux - WSL conveniently manages to connect them together is all! I might just move over to Linux proper and ditch the windows application and call it a day, or do some funky TKinter things to deliver what they want. 

Anyhow thanks for looking into it!

**Comment by Kalmat** [comment]
One last thing, just in case it is helpful: If you just want to control your own application windows, you can do so by instantiating pywinctl.Window(xid). You can get the xid value for your target window by using PyQt's winId() or Tkinter's root.frame() methods...

Thank YOU for your feedback!

**Comment by ubaldot** [comment]
I am one of the few (or many ehehhe) that is constrained to WSL. I was wondering if it is possible to make it working natively for WSL. Perhaps one idea is to call the function with something like `powershell.exe -c "python 3 ... " `?

**Comment by Kalmat** [comment]
Hi @ubaldot! Thank you for your reply!

Let me please try to explain the issue and the problems I found when trying to resolve it:
1. PyWinCtl requires an X-Server and a Window/Desktop Manager compatible with EWMH specs
2. WSL has none of them
3. WSL uses Wayland, which is not compliant with EWMH
4. Wayland has no way (using EWMH nor any other alternative) to return the list of open windows neither the active window when it is queried. This is not a bug or a missing feature, but a security measure they want to keep on purpose
5. Because of that, you have no access to "foreign" windows (I mean, windows from other applications apart from yours)
6. So, PyWinCtl will only work natively with your own application windows, which is not very useful, to be honest

In short: You can manage the windows from outside WSL (MS-Windows OS, I mean), but it doesn't seem to be possible in a native way. 

If you are interested or just curious, and you are skilled in WSL, you can try to install an X-Server to test, but I think that it will not be enough, since you may need to install a Desktop/Window Manager (e.g. GNOME) as well... I guess that's not your target architecture for your WSL projects.

Anyway, if any of you find a possible solution, I will be really happy to help!

**Comment by ubaldot** [comment]
That is crystal clear! :) 

What I had in mind is to "cheat" in the sense of calling powershell.exe from WSL2.
For example, consider the following example:

```
# pywinctl_example.py

import pywinctl as gw

# List all open windows
all_windows = gw.getAllWindows()

# Get a specific window by title
search_string = "example.txt"  # Replace with the title of the window you want to interact with
matching_windows = [
    window for window in all_windows if search_string in window.title
]

for window in matching_windows:
    # Activate the window
    window.activate()
    # Move the window to (100, 100)
    window.moveTo(100, 100)
    # Resize the window to 800x600
    window.resizeTo(800, 600)
```

From WSL2, instead of calling `python.exe -m pywinctl_example`,  you could invoke something like:

`powershell.exe -Command "cd \\\\wsl.localhost\Ubuntu-22.04\home\myname\tmp; python.exe -m pywinctl_example"`

assuming that you have a file called `example.txt` opened in WSL2 and that the script `~/tmp/pywinctl_example.py` exists in WSL.

I give it a quick shot and it work-ish, but then the "moved" window does not react at all.
